### PR TITLE
ARROW-866: [Python] Normalize PyErr exc_value to be more predictable

### PR DESCRIPTION
**Pull request by BryanCutler**
It is possible when using `PyErr_Fetch(&exc_type, &exc_value, &traceback)` for the `exc_value` to be a string, tuple or NULL.  Calling `PyErr_Normalize` after this will cause `exc_value` to always be a valid object of the same type as `exc_type` which can then be converted to a string predictably.

### CODE REVIEW

**Comment by BryanCutler** [comment]
cc @wesm I tested this again and seems to work fine

File: cpp/src/arrow/python/common.cc [hunk 1]
```diff
@@ -68,20 +68,16 @@ Status CheckPyError(StatusCode code) {
   if (PyErr_Occurred()) {
     PyObject *exc_type, *exc_value, *traceback;
     PyErr_Fetch(&exc_type, &exc_value, &traceback);
-    PyObjectStringify stringified(exc_value);
+    PyErr_NormalizeException(&exc_type, &exc_value, &traceback);
+    PyObject *exc_value_str = PyObject_Str(exc_value);
+    char *c_str = PyString_AsString(exc_value_str);
```
Comment (wesm) [inline cpp/src/arrow/python/common.cc]: Does this have unicode issues on Python 3? It may be safer to use `PyObjectStringify` on `exc_value_str`.

**Review by wesm — COMMENTED** [review]
Cool, thanks for digging into this gnarly issue. Minor comment

File: cpp/src/arrow/python/common.cc (hunk 1 above)
Comment (BryanCutler) [inline cpp/src/arrow/python/common.cc]: Yeah, I didn't notice that was only a Python2 function, let me try it again with PyObjectStringify

**Comment by BryanCutler** [comment]
Ok, it's working now with `PyObjectStringify` instead @wesm

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**